File: flutter_app/lib/ai/engine/body_analyzer.dart
import 'package:flutter/foundation.dart';
import 'dart:ui';
import 'dart:math';
import 'package:google_mlkit_pose_detection/google_mlkit_pose_detection.dart';
import '../models/ai_output.dart';
import 'pose_detector.dart';
import 'quality_checker.dart';
import 'feature_extractor.dart';
import 'estimators.dart';
import 'recommender.dart';

class BodyAnalyzer {
  final _poseDetector = PoseDetectorWrapper();
  final _qualityChecker = QualityChecker();
  final _featureExtractor = FeatureExtractor();
  final _estimators = Estimators();
  final _recommender = Recommender();

  Future<AiOutput> analyze({
    required InputImage inputImage,
    required Size imageSize, // Need dimensions
    required double heightCm,
    required double weightKg,
    required int age,
    required String gender,
    required String goal,
  }) async {
    // 1. Detect Pose
    final landmarks = await _poseDetector.detect(inputImage);
    
    // Default error structure
    AiOutput errorResponse(String msg, {String risk = "Risk", double poseScore = 0.0, Map<String, dynamic>? qualityChecks}) {
        return AiOutput(
            landmarksDetected: false,
            riskLevel: risk,
            recommendations: AiRecommendations(dailyCalories: 0, macros: {}, dietPlan: [], workoutPlan: []),
            debug: AiDebug(
                poseScore: poseScore,
                qualityChecks: qualityChecks ?? {'fullBodyVisible': false, 'poseOk': false, 'error': msg},
                rawRatios: {},
                error: msg
            )
        );
    }

    if (landmarks == null || landmarks.isEmpty) {
        debugPrint("AI_DEBUG: BodyAnalyzer - No human pose detected");
        return errorResponse("No human pose detected");
    }
    debugPrint("AI_DEBUG: BodyAnalyzer - Landmarks detected: ${landmarks.length}");
    
    // Summary of landmarks to verify uniqueness
    final first5 = landmarks.take(5).map((l) => "(${l.x.toInt()},${l.y.toInt()})").join(", ");
    debugPrint("AI_DEBUG: BodyAnalyzer - First 5 Landmarks: $first5");
    
    final minX = landmarks.map((l) => l.x).reduce(min);
    final maxX = landmarks.map((l) => l.x).reduce(max);
    final minY = landmarks.map((l) => l.y).reduce(min);
    final maxY = landmarks.map((l) => l.y).reduce(max);
    debugPrint("AI_DEBUG: BodyAnalyzer - BBox: [${minX.toInt()}, ${minY.toInt()}] to [${maxX.toInt()}, ${maxY.toInt()}]");

    // 2. BMI Baseline Fallback (Always have a result)
    final heightM = heightCm / 100;
    double bmiBaseline = weightKg / (heightM * heightM);
    
    // 3. Quality Check
    final qualityChecks = _qualityChecker.check(landmarks, imageSize);
    bool fullBodyVisible = qualityChecks['fullBodyVisible'] ?? false;
    bool poseOk = qualityChecks['poseOk'] ?? false;
    
    if (!fullBodyVisible || !poseOk) {
        debugPrint("AI_DEBUG: BodyAnalyzer - Quality Issues Detected (ABORTING ANALYSIS)");
        debugPrint("   > Visible: $fullBodyVisible, PoseOk: $poseOk");
        debugPrint("   > Messages: ${qualityChecks['messages']}");
        return errorResponse("Fotoğraf analiz için uygun değil. Lütfen tam vücut ve talimatlara uygun bir fotoğraf yükleyin.", qualityChecks: qualityChecks);
    }

    // 3. Features
    final features = _featureExtractor.extract(landmarks);

    // Score heuristic (average visibility of key points)
    double score = landmarks.map((e) => e.likelihood).reduce((a, b) => a + b) / landmarks.length;
    double confidence = (qualityChecks['confidenceScore'] as double?) ?? score;

    // 4. Estimates
    final estimates = _estimators.estimate(
        heightCm: heightCm,
        weightKg: weightKg,
        age: age,
        gender: gender,
        features: features,
        confidenceScore: confidence
    );

    // 5. Recommendations
    final recs = _recommender.recommend(
        age: age,
        gender: gender,
        weightKg: weightKg,
        heightCm: heightCm,
        goal: goal,
        riskLevel: estimates['riskLevel'],
        bodyFatPct: estimates['bodyFatPct'],
        leanMassKg: estimates['leanMassKg']
    );
    
    debugPrint("AI_DEBUG: BodyAnalyzer - Pose Score: $score, Final Confidence: $confidence");
    debugPrint("AI_DEBUG: BodyAnalyzer - Estimates: $estimates");

    return AiOutput(
        landmarksDetected: true,
        bmi: estimates['bmi'],
        bodyFatPct: estimates['bodyFatPct'],
        leanMassKg: estimates['leanMassKg'],
        confidenceScore: confidence,
        riskLevel: estimates['riskLevel'],
        recommendations: recs,
        debug: AiDebug(
            poseScore: score,
            qualityChecks: qualityChecks,
            rawRatios: features,
        )
    );
  }

  void dispose() {
    _poseDetector.close();
  }
}

File: backend/ai_engine/body_analyzer.py
from .pose_detector import PoseDetectorWrapper
from .quality_checker import QualityChecker
from .feature_extractor import FeatureExtractor
from .estimators import Estimators
from .recommender import Recommender

class BodyAnalyzer:
    def __init__(self):
        self._pose_detector = PoseDetectorWrapper()
        self._quality_checker = QualityChecker()
        self._feature_extractor = FeatureExtractor()
        self._estimators = Estimators()
        self._recommender = Recommender()

    def analyze(self, image, height_cm, weight_kg, age, gender, goal):
        """
        Analyzes the image and user stats.
        Returns a dictionary with success boolean and either an error or the full analysis.
        """
        # 1. Detect Pose
        landmarks = self._pose_detector.detect(image)
        
        def error_response(msg, risk="Risk", pose_score=0.0, quality_checks=None):
            if quality_checks is None:
                quality_checks = {'fullBodyVisible': False, 'poseOk': False, 'error': msg}
            return {
                "success": False,
                "error": "invalid_photo", # This matches Flutter expectations
                "message": msg,
                "debug": {
                    "poseScore": pose_score,
                    "qualityChecks": quality_checks
                }
            }

        if not landmarks:
            print("AI_DEBUG: BodyAnalyzer - No human pose detected")
            return error_response("No human pose detected")
            
        print(f"AI_DEBUG: BodyAnalyzer - Landmarks detected: {len(landmarks)}")

        # Print first 5
        first5 = ", ".join([f"({int(l.x*image.shape[1])},{int(l.y*image.shape[0])})" for l in landmarks[:5]])
        print(f"AI_DEBUG: BodyAnalyzer - First 5 Landmarks: {first5}")

        x_coords = [l.x * image.shape[1] for l in landmarks]
        y_coords = [l.y * image.shape[0] for l in landmarks]
        print(f"AI_DEBUG: BodyAnalyzer - BBox: [{int(min(x_coords))}, {int(min(y_coords))}] to [{int(max(x_coords))}, {int(max(y_coords))}]")

        # 2. Quality Check
        # image.shape is (height, width, channels)
        image_size = (image.shape[1], image.shape[0]) 
        quality_checks = self._quality_checker.check(landmarks, image_size)
        
        full_body_visible = quality_checks.get('fullBodyVisible', False)
        pose_ok = quality_checks.get('poseOk', False)

        if not full_body_visible or not pose_ok:
            print("AI_DEBUG: BodyAnalyzer - Quality Issues Detected (ABORTING ANALYSIS)")
            print(f"   > Visible: {full_body_visible}, PoseOk: {pose_ok}")
            print(f"   > Messages: {quality_checks.get('messages')}")
            return error_response("Fotoğraf analiz için uygun değil. Lütfen tam vücut ve talimatlara uygun bir fotoğraf yükleyin.", quality_checks=quality_checks)

        # 3. Features
        features = self._feature_extractor.extract(landmarks, image_size)

        # Score heuristic
        score = sum([l.visibility for l in landmarks]) / len(landmarks)
        confidence = quality_checks.get('confidenceScore', score)

        # 4. Estimates
        estimates = self._estimators.estimate(
            height_cm=height_cm,
            weight_kg=weight_kg,
            age=age,
            gender=gender,
            features=features,
            confidence_score=confidence
        )

        # 5. Recommendations
        recs = self._recommender.recommend(
            age=age,
            gender=gender,
            weight_kg=weight_kg,
            height_cm=height_cm,
            goal=goal,
            risk_level=estimates['riskLevel'],
            body_fat_pct=estimates['bodyFatPct'],
            lean_mass_kg=estimates['leanMassKg']
        )
        
        print(f"AI_DEBUG: BodyAnalyzer - Pose Score: {score}, Final Confidence: {confidence}")
        print(f"AI_DEBUG: BodyAnalyzer - Estimates: {estimates}")

        return {
            "success": True,
            "landmarksDetected": True,
            "bmi": estimates['bmi'],
            "bodyFatPct": estimates['bodyFatPct'],
            "leanMassKg": estimates['leanMassKg'],
            "confidenceScore": confidence,
            "riskLevel": estimates['riskLevel'],
            "calories": recs['dailyCalories'],
            "macros": recs['macros'],
            "dietPlan": recs['dietPlan'],
            "workoutPlan": recs['workoutPlan'],
            "debug": {
                "poseScore": score,
                "qualityChecks": quality_checks,
                "rawRatios": features
            }
        }

    def close(self):
        self._pose_detector.close()
